### PR TITLE
alwaysIdentifier annotation option implementation

### PR DESCRIPTION
**Pull request by adelplace**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1130 
| License       | MIT
| Doc PR        | https://github.com/api-platform/docs/pull/373

The goal here is to allow the users to display only the IRI of a nested resource. It's very usefull on trees or on big entities which have multiple links on a same other entity.

As the MaxDepth annotation doesn't allow us to do that I have created a new annotation option :
```php
/**
 * @ApiResource(attributes={
 *     "normalization_context"={"groups": {"alwaysIdentifierGroup"}, "enable_always_identifier"=true}
 * })
 * @ORM\Entity
 */
class AlwaysIdentifierDummy 
{
    /**
     * @var ArrayCollection children dummies
     *
     * @ORM\ManyToMany(targetEntity="AlwaysIdentifierDummy")
     * @ApiProperty(alwaysIdentifier=true)
     * @Groups({"alwaysIdentifierGroup"})
     */
    private $children;
}
```

Will be displayed as :

```json
{
      "children": [
          "/always_identifier_dummies/1"
      ]
}
```

I don't know if this work fit with the project guidelines but i would be happy to discuss about it ;)

### CODE REVIEW

**Comment by soyuka** [comment]
We're successfully doing this by using groups and max depth.

**Comment by adelplace** [comment]
> What about using an attribute of @ApiProperty instead of a new annotation? For example @ApiProperty(skip_serialization=true) (not sure about the key).

Yes it would surely be a better solution.

> I'm not fond of the word IriOnly by the way because we have to keep in mind that IRI's are related to JSON-ld and sometimes we might just want id's properties (ie force to skip serialization of the property).

I'll look to implement it on the other formats.

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -81,6 +88,18 @@ public function normalize($object, $format = null, array $context = [])
...
     }
 
+    /**
+     * Is IRI only enabled on this item ?
+     *
+     * @param array $context
+     *
+     * @return bool
+     */
+    private function isIriOnlyItem(array &$context)
```
Comment (meyerbaptiste) [inline src/JsonLd/Serializer/ItemNormalizer.php]: Why is `$context` passed by reference?

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -89,6 +108,35 @@ public function supportsDenormalization($data, $type, $format = null)
...
 
+    /**
+     * Is Iri Only enabled for this attribute ?
+     *
+     * @param $propertyMetadata
+     * @param array $context
+     *
+     * @return bool
+     */
+    private function isIriOnlyAttribute($propertyMetadata, array &$context)
```
Comment (meyerbaptiste) [inline src/JsonLd/Serializer/ItemNormalizer.php]: Why is `$context` passed by reference?

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 3]
```diff
@@ -89,6 +108,35 @@ public function supportsDenormalization($data, $type, $format = null)
...
+    {
+        return
+            isset($context[static::ENABLE_IRI_ONLY]) &&
```
Comment (meyerbaptiste) [inline src/JsonLd/Serializer/ItemNormalizer.php]: This class is final, use `self` instead of `static`.

**Comment by adelplace** [comment]
@soyuka @meyerbaptiste 
The changes have been done, tell me if it's better. I also added tests for json and hal formats.

File: src/Metadata/Property/Factory/AnnotationPropertyMetadataFactory.php [hunk 4]
```diff
@@ -117,12 +117,14 @@ private function createMetadata(ApiProperty $annotation, PropertyMetadata $paren
                 $annotation->identifier,
                 $annotation->iri,
                 null,
-                $annotation->attributes
+                $annotation->attributes,
+                null,
+                $annotation->skipSerialization
```
Comment (soyuka) [inline src/Metadata/Property/Factory/AnnotationPropertyMetadataFactory.php]: I think that the word "serializable" would be better:

```
isSerializable
withSerializable
serializable
```

File: src/Serializer/AbstractItemNormalizer.php [hunk 5]
```diff
@@ -387,6 +406,21 @@ protected function createRelationSerializationContext(string $resourceClass, arr
...
 
+    /**
+     * Is normalization skipping enabled for this attribute ?
+     *
+     * @param $propertyMetadata
+     * @param array $context
+     *
+     * @return bool
+     */
+    private function isSkippedSerializationAttribute($propertyMetadata, array $context)
```
Comment (soyuka) [inline src/Serializer/AbstractItemNormalizer.php]: `isPropertySerializable`

Also missing typehints and you can remove the block comments (`@param/@return`):

```
function isPropertySerializable(PropertyMetadata $propertyMetadata, array $context): bool 
{}
```
Then negate: `!$this->isPropertySerializable(...)`

File: src/Serializer/AbstractItemNormalizer.php [hunk 6]
```diff
@@ -97,9 +100,25 @@ public function normalize($object, $format = null, array $context = [])
...
     }
 
+    /**
+     * Is normalization should be skipped on this item ?
+     *
+     * @param array $context
+     *
+     * @return bool
+     */
+    protected function isItemSerializationShouldBeSkipped(array $context)
```
Comment (soyuka) [inline src/Serializer/AbstractItemNormalizer.php]: `shouldSkipSerialization` same you can add a return type `:bool` and remove `@param/@return`.

**Comment by dunglas** [comment]
What about `always_serialize_as_id` instead? I don't like the "skip serialization" term because the serialization is not skipped (to me it would mean "just skip this property", here it is serialized in a special type: an IRI (btw, IRI is a generic term; not specific to JSON-LD: https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier

**Comment by soyuka** [comment]
👍 @dunglas , I'd propose in the metadata:

```
isAlwaysIdentifier
withAlwaysIdentifier
alwaysIdentifier
```

**Comment by adelplace** [comment]
@soyuka  @dunglas

File: features/main/skip_serialization.feature [hunk 7]
```diff
@@ -0,0 +1,118 @@
+@skip
```
Comment (soyuka) [inline features/main/skip_serialization.feature]: why skip?

File: src/Metadata/Property/PropertyMetadata.php [hunk 8]
```diff
@@ -381,4 +383,29 @@ public function withSubresource(SubresourceMetadata $subresource = null): self
 
         return $metadata;
     }
+
+    /**
+     * Returns a new instance with the given alwaysIdentifier flag.
+     *
+     * @param bool $alwaysIdentifier
```
Comment (dunglas) [inline src/Metadata/Property/PropertyMetadata.php]: You can remove both `@param` and `@return`, they are useless because of the typehints.

File: src/Serializer/AbstractItemNormalizer.php [hunk 9]
```diff
@@ -97,9 +100,21 @@ public function normalize($object, $format = null, array $context = [])
             $context['resources'][$resource] = $resource;
         }
 
+        if ($this->isAlwaysIdentifierItem($context)) {
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]: Can you inline this one? We try to reduce the number of function calls in the serialization process (which is recursive) because a function call is costly:

```php
if (($context[self::ALWAYS_IDENTIFIER_KEY] ?? false) && ($context[self::ALWAYS_IDENTIFIER_KEY] ?? false))
```

Same for `isAlwaysIdentifierProperty`, I would drop these methods.

File: src/Serializer/AbstractItemNormalizer.php [hunk 10]
```diff
@@ -39,6 +39,9 @@
 {
     use ContextTrait;
 
+    const ENABLE_ALWAYS_IDENTIFIER = 'enable_always_identifier';
+    const ALWAYS_IDENTIFIER_KEY = 'always_identifier_key';
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]: We use the `_key` suffix nowhere. Can you drop it?

File: tests/Fixtures/TestBundle/Entity/AlwaysIdentifierDummy.php [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Serializer\Annotation\Groups;
+
+/**
+ * AlwaysIdentifierDummy.
+ *
+ * @author Alexandre Delplace <alexandre.delplacemille@gmail.com>
+ *
+ * @ApiResource(attributes={
+ *     "normalization_context"={"groups"={"alwaysIdentifierGroup"}, "enable_always_identifier"=true}
```
Comment (dunglas) [inline tests/Fixtures/TestBundle/Entity/AlwaysIdentifierDummy.php]: Why do we need this `enable_always_identifier` flag? IMO we can remove it, if the `alwaysIdentifier` is set on a property, we use it, that's all, it doesn't make sense to allow to disable this feature. If you don't want to use it, just don't set the flag.

File: features/main/always_identifier.feature [hunk 12]
```diff
@@ -0,0 +1,117 @@
+Feature: AlwaysIdentifier support
```
Comment (dunglas) [inline features/main/always_identifier.feature]: May  I suggest: `Feature: allow resources in properties to always be serialized as identifiers`

File: features/main/always_identifier.feature [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  In order to use a hypermedia API
+  As a client software developer
+  I need to be able to force identifier display of related entities
```
Comment (dunglas) [inline features/main/always_identifier.feature]: I need to be able to force to serialize only the identifier of the related resource

File: features/main/always_identifier.feature [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+  @createSchema
+  Scenario: Create a set of dummy with alwaysIdentifier attributes and check the jsonld formatted response
```
Comment (dunglas) [inline features/main/always_identifier.feature]: Create a set of Dummy with the alwaysIdentifier attributes and check the JSON-LD formatted response

File: features/main/always_identifier.feature [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+                "children": [],
+                "related": [],
+                "parent": null
+            }
+        ],
+        "parent": "/always_identifier_dummies/1"
+    }
+    """
+
+    Scenario: Check the json formatted response
```
Comment (dunglas) [inline features/main/always_identifier.feature]: JSON

File: features/main/always_identifier.feature [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+      When I add "Accept" header equal to "application/json"
+      And I add "Content-Type" header equal to "application/json"
+      And I send a "GET" request to "/always_identifier_dummies/2"
+      Then the JSON should be equal to:
+      """
+      {
+          "children": [
+              "\/always_identifier_dummies\/1"
```
Comment (dunglas) [inline features/main/always_identifier.feature]: Can you drop the `\` please (for readability), same for all others.

File: features/main/always_identifier.feature [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+                  "children": [],
+                  "related": [],
+                  "parent": null
+              }
+          ],
+          "parent": "\/always_identifier_dummies\/1"
+      }
+      """
+
+  Scenario: Check the hal formatted response
```
Comment (dunglas) [inline features/main/always_identifier.feature]: HAL

File: features/main/always_identifier.feature [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+                      }
+                  }
+              ],
+              "parent": "\/always_identifier_dummies\/1"
+          }
+      }
+      """
+
+  @dropSchema
+  Scenario: Delete a resource
```
Comment (dunglas) [inline features/main/always_identifier.feature]: This is not necessary (this is not related to this feature and already tested). `@dropSchema` will already destroy the database.

File: features/main/always_identifier.feature [hunk 19]
```diff
@@ -0,0 +1,117 @@
+Feature: AlwaysIdentifier support
+  In order to use a hypermedia API
```
Comment (dunglas) [inline features/main/always_identifier.feature]: You can create a new context method in `FeatureContext` like https://github.com/api-platform/core/blob/master/features/bootstrap/FeatureContext.php#L125 to create new `AlwaysIdentifiersDummies`, remove the create and delete scenarios, and move the scenarios for each format (JSON-LD, HAL, JSON) in the corresponding directories (we prefer to group tests per format instead of per feature, it will help when we will create a subtree split).

File: tests/Fixtures/TestBundle/Entity/AlwaysIdentifierDummy.php (hunk 11 above)
Comment (adelplace) [reply tests/Fixtures/TestBundle/Entity/AlwaysIdentifierDummy.php]: I took example from the max_depth implementation. I'll remove it

**Comment by adelplace** [comment]
@dunglas done. I also added a json folder in the features because I didn't find a specific one for this format.

**Comment by dunglas** [comment]
Thanks @adelplace
